feat: raise an exception if workflows have zero tasks.

BenjaminMummery opened the pull request:
[ORQSDK-774]

# The problem

On QE you get an HTTP error stating that you need at least 1 task to be executed if your WF doesn't use any tasks. This is not the case on local Ray or CE. We should ensure that this is consistent across all runtimes.

# This PR's solution

- Adds a check during traversal that raises a WorkflowSyntaxError when no task invocations are identified. 
  - The exception reads: "The workflow '[wf name]' cannot be submitted as it does not define any tasks to be executed. Please modify the workflow definition to define at least one task and retry submitting the workflow."
  - If the source file and line can be identified, these are also included in the message.

# Checklist

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`


[ORQSDK-774]: https://zapatacomputing.atlassian.net/browse/ORQSDK-774?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

BenjaminMummery committed:
feat: raise an exception if workflows have zero tasks.

[ORQSDK-774]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-774-warn-user-when-wf-does-not-have-a-single-task

BenjaminMummery committed:
test: confirm that 0-task workflow raises an exception.

[ORQSDK-774]

BenjaminMummery committed:
fix: acutally test

[ORQSDK-774]

SebastianMorawiec commented on src/orquestra/sdk/_base/_workflow.py:
```diff
@@ -113,7 +115,14 @@ def model(self) -> ir.WorkflowDef:
+
+        if len(model.task_invocations) < 1:
+            raise WorkflowSyntaxError(
+                f"The workflow '{model.name}' requires 0 tasks to be executed. "
```
Suggestion: TBH I was a bit confused at first by this text. Maybe something more straight user-facing? Like "Your wofklow definition (or the workflow you are trying to submit? Whatever works better for you) [...]." 
But this thing here works for me as well.

BenjaminMummery replied:
Good thought, thanks! [b383015](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/86/commits/b383015b8e90c27bae6045f6859dfd86de99141e) updates this to be a little more user-facing and also include the file and line number where the definition is found to help point the user to the changes they need to make. It now reads:

```
"The workflow '[wf name]' (defined at [file] line [line]) cannot be submitted as it does not define any tasks to be executed. Please modify the workflow definition to define at least one task and retry submitting the workflow."
```

SebastianMorawiec reviewed (approved):
👍

BenjaminMummery committed:
feat: make help message for missing tasks more user-facing.

[ORQSDK-774]
Addresses https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/86/files#r1141811731

BenjaminMummery committed:
docs: update changelog with >0 tasks requirement.

[ORQSDK-774]

BenjaminMummery committed:
style: remove assumption that model.fn_ref necessarily has file_path and line_number attrs.

[ORQSDK-774]
